Headers/DebugServer2/Host/Windows/ExtraWrappers.h

```diff
@@ -358,23 +357,80 @@ BOOL WINAPI VirtualFreeEx(
   _In_ DWORD  dwFreeType
 );
 
-DWORD WINAPI GetModuleFileNameExA(
-  _In_     HANDLE  hProcess,
-  _In_opt_ HMODULE hModule,
-  _Out_    LPSTR   lpFilename,
-  _In_     DWORD   nSize
+BOOL WINAPI DebugBreakProcess(
+  _In_ HANDLE Process
 );
-DWORD WINAPI GetModuleFileNameExW(
-  _In_     HANDLE  hProcess,
-  _In_opt_ HMODULE hModule,
-  _Out_    LPWSTR  lpFilename,
-  _In_     DWORD   nSize
+
+#define TH32CS_SNAPTHREAD 0x00000004
+HANDLE WINAPI CreateToolhelp32Snapshot(
+  _In_  DWORD dwFlags,
+  _In_  DWORD th32ProcessID
 );
-#define GetModuleFileNameEx GetModuleFileNameExW
 
+typedef struct tagTHREADENTRY32 {
+  DWORD dwSize;
+  DWORD cntUsage;
+  DWORD th32ThreadID;
+  DWORD th32OwnerProcessID;
+  LONG  tpBasePri;
+  LONG  tpDeltaPri;
+  DWORD dwFlags;
+} THREADENTRY32, *PTHREADENTRY32, *LPTHREADENTRY32;
+BOOL WINAPI Thread32First(
+  _In_     HANDLE hSnapshot,
+  _Inout_  LPTHREADENTRY32 lpte
+);
+
+BOOL WINAPI Thread32Next(
+  _In_   HANDLE hSnapshot,
+  _Out_  LPTHREADENTRY32 lpte
+);
+
+} // extern "C"
+// clang-format on
+
+// The following code allows us to pick some symbols directly from kernel32.dll
+// (against which we cannot link when building for WinStore-ARM for instance.
+// The functions are declared in the above block (see DebugBreakProcess for
+// instance), but these declarations are then hidden with macros below.
+// Basically, we just
+//     GetProcAddress(GetModuleHandle("kernel32"), funcName)
+// and call that.
+//
+// Given that we have declarations for these functions above, we can also
+// type-check these calls and make sure both the return type and the arguments
+// type (see `decltype` in `DO_K32_CALL`) are matching.
+//
+// If we need more functions that are only available in kernel32.dll, we just
+// need to add their prototype in the above `extern "C" {}` block, and add a
+// one line `#define` below, in this form:
+//     #define FunctionName(...) DO_K32_CALL(FunctionName, __VA_ARGS__)
+
+static inline FARPROC GetK32Proc(_In_ LPCSTR procName) {
+  static HMODULE kernel32Base = nullptr;
+  if (kernel32Base == nullptr) {
+    kernel32Base = GetModuleHandle(L"kernel32");
```

Comment:
There's only one thread in ds2 that can call this function. I think if
there was a race for the initialization, we could call `GetModuleHandle`
twice instead of just once, but there still wouldn't be a bug cause it's
just a pointer and it should be the same for every call.

Thanks for pointing it out though, I hadn't thought about it.

On Mon, Mar 21, 2016 at 2:15 PM, Dave Lee <notifications@github.com> wrote:

> In Headers/DebugServer2/Host/Windows/ExtraWrappers.h
> <https://github.com/facebook/ds2/pull/235#discussion_r56898808>:
>
> > +//     GetProcAddress(GetModuleHandle("kernel32"), funcName)
> > +// and call that.
> > +//
> > +// Given that we have declarations for these functions above, we can also
> > +// type-check these calls and make sure both the return type and the arguments
> > +// type (see `decltype` in `DO_K32_CALL`) are matching.
> > +//
> > +// If we need more functions that are only available in kernel32.dll, we just
> > +// need to add their prototype in the above `extern "C" {}` block, and add a
> > +// one line `#define` below, in this form:
> > +//     #define FunctionName(...) DO_K32_CALL(FunctionName, __VA_ARGS__)
> > +
> > +static inline FARPROC GetK32Proc(_In_ LPCSTR procName) {
> > +  static HMODULE kernel32Base = nullptr;
> > +  if (kernel32Base == nullptr) {
> > +    kernel32Base = GetModuleHandle(L"kernel32");
>
> Is there any chance of the double initialization race here being a problem?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/ds2/pull/235/files/817838778e518fff191ed2166a85717de8295c94#r56898808>
>



-- 
Stephane Sezer